-1

**Failed Tests**:   Build  HeaderConsistency  ClangBuild 
**Size**: This PR adds an extra 84KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d8e0dd/42244/summary.html
**COMMIT**: 09cac6a1300e31e0a3e210db702a5c0cff40346d
**CMSSW**: CMSSW_14_2_X_2024-10-15-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46411/42244/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @bsunanda cms-sw/cmssw#46386

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d8e0dd/42244/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d8e0dd/42244/git-merge-result


## Build

I found compilation error when building: 

<pre>In file included from src/L1Trigger/L1TMuonOverlap/interface/OMTFSorter.h:12,
                 from src/L1Trigger/L1TMuonOverlap/interface/OMTFReconstruction.h:23,
                 from src/L1Trigger/L1TMuonOverlap/plugins/L1TMuonOverlapTrackProducer.h:19,
                 from src/L1Trigger/L1TMuonOverlap/plugins/L1TMuonOverlapTrackProducer.cc:10:
src/L1Trigger/L1TMuonOverlap/interface/AlgoMuon.h: In constructor 'AlgoMuon::AlgoMuon(int, int, int, int, int, int, int, int, int)':
src/L1Trigger/L1TMuonOverlap/interface/AlgoMuon.h:90:16: error: 'AlgoMuon::m_rhitNumb' will be initialized after [-Werror=reorder]
   90 |   unsigned int m_rhitNumb;
      |                ^~~~~~~~~~
src/L1Trigger/L1TMuonOverlap/interface/AlgoMuon.h:86:7: error:   'int AlgoMuon::m_phiRHit' [-Werror=reorder]
   86 |   int m_phiRHit;
      |       ^~~~~~~~~
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Entering Package L1Trigger/TrackFindingTMTT
>> Entering Package L1Trigger/TrackFindingTracklet
>> Entering Package L1Trigger/TrackTrigger
>> Entering Package L1TriggerConfig/L1GtConfigProducers
>> Compile sequence completed for CMSSW CMSSW_14_2_X_2024-10-15-2300
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
+ eval scram build outputlog '&&' '(python3' /data/cmsbld/jenkins/workspace/ib-run-pr-tests/cms-bot/buildLogAnalyzer.py --logDir /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_2_X_2024-10-15-2300/tmp/el8_amd64_gcc12/cache/log/src '||' 'true)'
++ scram build outputlog
>> Entering Package CondTools/Geometry
Entering library rule at src/CondTools/Geometry/plugins
>> Compiling edm plugin src/CondTools/Geometry/plugins/CSCRecoIdealDBLoader.cc
</pre>

--- a/L1Trigger/DTTriggerPhase2/src/MuonPathAssociator.cc
+++ b/L1Trigger/DTTriggerPhase2/src/MuonPathAssociator.cc
@@ -589,7 +589,7 @@ void MuonPathAssociator::correlateMPaths(edm::Handle<DTDigiCollection> dtdigis,
         sl3 = 0;
         for (auto SL3metaPrimitive = SL3metaPrimitives.begin(); SL3metaPrimitive != SL3metaPrimitives.end();
              ++SL3metaPrimitive, sl3++) {
-          if (useFitSL3[sl3])
+          [[clang::suppress]] if (useFitSL3[sl3])
             continue;
           if ((at_least_one_correlation == false || clean_chi2_correlation_) &&
               allow_confirmation_) {  //no correlation was found, trying with pairs of two digis in the other SL
@@ -841,7 +841,7 @@ void MuonPathAssociator::correlateMPaths(edm::Handle<DTDigiCollection> dtdigis,
             sl1 = 0;
             for (auto SL1metaPrimitive = SL1metaPrimitives.begin(); SL1metaPrimitive != SL1metaPrimitives.end();
                  ++SL1metaPrimitive, sl1++) {
-              if (useFitSL1[sl1])
+              [[clang::suppress]] if (useFitSL1[sl1])
                 continue;
 
               DTSuperLayerId SLId(SL1metaPrimitive->rawId);
@@ -902,7 +902,7 @@ void MuonPathAssociator::correlateMPaths(edm::Handle<DTDigiCollection> dtdigis,
             sl3 = 0;
             for (auto SL3metaPrimitive = SL3metaPrimitives.begin(); SL3metaPrimitive != SL3metaPrimitives.end();
                  ++SL3metaPrimitive, sl3++) {
-              if (useFitSL3[sl3])
+              [[clang::suppress]] if (useFitSL3[sl3])
                 continue;
               DTSuperLayerId SLId(SL3metaPrimitive->rawId);
               DTChamberId(SLId.wheel(), SLId.station(), SLId.sector());
@@ -1008,12 +1008,12 @@ void MuonPathAssociator::removeSharingFits(vector<metaPrimitive> &chamberMPaths,
   for (unsigned int i = 0; i < chamberMPaths.size(); i++) {
     if (debug_)
       LogDebug("MuonPathAssociator") << "Looking at prim" << i;
-    if (!useFit[i])
+    [[clang::suppress]] if (!useFit[i])
       continue;
     for (unsigned int j = i + 1; j < chamberMPaths.size(); j++) {
       if (debug_)
         LogDebug("MuonPathAssociator") << "Comparing with prim " << j;
-      if (!useFit[j])
+      [[clang::suppress]] if (!useFit[j])
         continue;
       metaPrimitive first = chamberMPaths[i];
       metaPrimitive second = chamberMPaths[j];

--- a/L1Trigger/DTTriggerServerPhi/src/DTTSPhi.cc
+++ b/L1Trigger/DTTriggerServerPhi/src/DTTSPhi.cc
@@ -374,9 +374,9 @@ void DTTSPhi::runTSPhi() {
                  ntsm[is - 1 - DTConfigTSPhi::NSTEPF][i_tsmd] >
                      0) {  // it means that the last BX with sort 2 was not the
                            // previous one
-        existSecondPrevBx =
-            ((is - 1 - DTConfigTSPhi::NSTEPF >= 0) && (ntsm[is - 1 - DTConfigTSPhi::NSTEPF][i_tsmd] > 1) &&
-             (secondPrevBx->tracoTr()->code() > 0));
+        existSecondPrevBx = secondPrevBx && ((is - 1 - DTConfigTSPhi::NSTEPF >= 0) &&
+                                             (ntsm[is - 1 - DTConfigTSPhi::NSTEPF][i_tsmd] > 1) &&
+                                             (secondPrevBx->tracoTr()->code() > 0));
         if (existSecondPrevBx) {
           _cache.push_back(DTChambPhSegm(ChamberId(), is - 1, secondPrevBx->tracoTr(), 2));
 

--- a/L1Trigger/L1TCalorimeter/plugins/L1TPhysicalEtAdder.cc
+++ b/L1Trigger/L1TCalorimeter/plugins/L1TPhysicalEtAdder.cc
@@ -12,6 +12,7 @@
 #include "DataFormats/L1TCalorimeter/interface/CaloRegion.h"
 
 #include "FWCore/Framework/interface/ESHandle.h"
+#include <cassert>
 #include <vector>
 
 namespace {
@@ -54,7 +55,7 @@ namespace {
                                      3.750,
                                      4.250,
                                      4.750};
-    if (etaIndex < 11) {
+    if (0 <= etaIndex && etaIndex < 11) {
       return -rgnEtaValues[-(etaIndex - 10)];  // 0-10 are negative eta values
     } else if (etaIndex < 22) {
       return rgnEtaValues[etaIndex - 11];  // 11-21 are positive eta values

--- a/L1Trigger/L1TCalorimeter/src/firmware/Stage1Layer2EtSumAlgorithmImpPP.cc
+++ b/L1Trigger/L1TCalorimeter/src/firmware/Stage1Layer2EtSumAlgorithmImpPP.cc
@@ -168,6 +168,7 @@ int l1t::Stage1Layer2EtSumAlgorithmImpPP::DiJetPhi(const std::vector<l1t::Jet>* 
 uint16_t l1t::Stage1Layer2EtSumAlgorithmImpPP::MHToverHT(uint16_t num, uint16_t den) const {
   uint16_t result;
   uint32_t numerator(num), denominator(den);
+  assert(den != 0);
 
   if (numerator == denominator)
     result = 0x7f;

--- a/L1Trigger/L1THGCal/src/HGCalVFECompressionImpl.cc
+++ b/L1Trigger/L1THGCal/src/HGCalVFECompressionImpl.cc
@@ -1,3 +1,5 @@
+#include <cassert>
+
 #include "L1Trigger/L1THGCal/interface/HGCalVFECompressionImpl.h"
 
 #include "FWCore/Utilities/interface/Exception.h"
@@ -40,6 +42,7 @@ void HGCalVFECompressionImpl::compressSingle(const uint64_t value,
 
   // build exponent and mantissa
   const uint32_t exponent = bitlen - mantissaBits_;
+  assert(exponent - 1 < 64);
   const uint64_t mantissa = (shifted_value >> (exponent - 1)) & ~(1ULL << mantissaBits_);
 
   // assemble floating-point

--- a/L1Trigger/L1THGCal/src/backend/HGCalClusteringImpl.cc
+++ b/L1Trigger/L1THGCal/src/backend/HGCalClusteringImpl.cc
@@ -84,7 +84,7 @@ void HGCalClusteringImpl::clusterizeDR(const std::vector<edm::Ptr<l1t::HGCalTrig
       }
     }
 
-    if (targetClu < 0 && isSeed[itc])
+    [[clang::suppress]] if (targetClu < 0 && isSeed[itc])
       clustersTmp.emplace_back(*tc);
     else if (targetClu >= 0)
       clustersTmp.at(targetClu).addConstituent(*tc);
@@ -298,7 +298,7 @@ void HGCalClusteringImpl::clusterizeDRNN(const std::vector<edm::Ptr<l1t::HGCalTr
     /* continue if not passing the threshold */
     if ((*tc)->mipPt() < threshold)
       continue;
-    if (isSeed[itc])
+    [[clang::suppress]] if (isSeed[itc])
       continue;  //No sharing of seeds between clusters (TBC)
 
     /* searching for TC near the centre of the cluster  */

--- a/L1Trigger/L1TMuonOverlap/interface/AlgoMuon.h
+++ b/L1Trigger/L1TMuonOverlap/interface/AlgoMuon.h
@@ -35,6 +35,7 @@ public:
         m_bx(bx),
         m_pt(pt),
         m_charge(charge),
+        m_phiRHit(9999),
         m_patNumb(999),
         m_rhitNumb(999) {}
 

